api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2294,6 +2297,200 @@ public void copyObject(String bucketName, String objectName, String destBucketNa
+   * @throws InsufficientDataException  upon getting EOFException while reading given
+   *           InputStream even before reading given length
+   * @throws IOException                 upon connection error
+   * @throws InvalidKeyException
+   *           upon an invalid access key or secret key
+   * @throws NoResponseException         upon no response from server
+   * @throws XmlPullParserException      upon parsing response xml
+   * @throws ErrorResponseException      upon unsuccessful execution
+   * @throws InternalException           upon internal library error
+   * @throws InvalidArgumentException    upon invalid value is passed to a method.
+   */
+
+
+  public void  composeObject(String bucketName, String objectName, ServerSideEncryption sse,
+                             Map<String,String> userMetaData, List<CopySource> copySources)
+          throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
+          InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+          InternalException, InvalidArgumentException {
+
+    if (copySources.size() == 0 || copySources.size() > MAX_MULTIPART_COUNT ) {
+      throw new InvalidArgumentException("There must be as least one and up to MAX_MULTIPART_COUNT source objects.");
+    }
+    int i = 1;
+    int totalParts = 0;
+    long totalSize = 0;
+    for (CopySource sourceObject : copySources) {
+
+      ObjectStat os = statObject(sourceObject.bucketName(), sourceObject.objectName());
+      // Error out if client side encryption is used in this source object when
+      // more than one source objects are given.
+      if ((copySources.size() > 1 )  && sourceObject.getHeaders().containsKey("x-amz-meta-x-amz-key")) {
+        throw new InvalidArgumentException("Client side encryption is used in source object"
+                + sourceObject.bucketName() + sourceObject.objectName());
+      }
+
+      // Check if a segment is specified is within object bounds
+      if ( sourceObject.length() != null && sourceObject.length() > os.length()) {
+        throw new InvalidArgumentException("Source specified length " + sourceObject.length() + " is more than  "
+                +  "size of Object " + os.length());
+      }
+
+      if ( os.length() < MIN_MULTIPART_SIZE && i < copySources.size() - 1 ) {
+        throw new InvalidArgumentException(" Source Object size " + os.length() + " is too small "
+         +  " and " + i + " it is not the last part");
+      }
+
+      totalSize = totalSize + os.length();
+      if ( totalSize > MAX_OBJECT_SIZE ) {
+        throw new InvalidArgumentException(" Cannot compose an object of size more than " + MAX_OBJECT_SIZE );
+      }
+      if ( sourceObject.length() != null ) {
+        totalParts += calculateMultipartSize(sourceObject.length())[1];
+      } else {
+        totalParts += calculateMultipartSize((int)os.length())[1];
+      }
```

Comment:
you use `os.length()` and `sourceObject.length()` based on conditions.  This makes logical error here and above.  Its better to define `length` variable and use `length` variable appropriately.